src/proto/network/pxe.rs

```diff
@@ -0,0 +1,1276 @@
+//! PXE Base Code protocol.
+
+use core::{ffi::c_void, iter::from_fn, ptr::NonNull};
+
+use bitflags::bitflags;
+use uefi_macros::{unsafe_guid, Protocol};
+
+use crate::{CStr8, Char8, Result, Status};
+
+use super::{IpAddress, MacAddress};
+
+/// PXE Base Code protocol
+#[repr(C)]
+#[unsafe_guid("03c4e603-ac28-11d3-9a2d-0090273fc14d")]
+#[derive(Protocol)]
+#[allow(clippy::type_complexity)]
+pub struct BaseCode {
+    revision: u64,
+    start: extern "efiapi" fn(this: &Self, use_ipv6: bool) -> Status,
+    stop: extern "efiapi" fn(this: &Self) -> Status,
+    dhcp: extern "efiapi" fn(this: &Self, sort_offers: bool) -> Status,
+    discover: extern "efiapi" fn(
+        this: &Self,
+        ty: BootstrapType,
+        layer: &mut u16,
+        use_bis: bool,
+        info: Option<*const DiscoverInfo<[Server; 0]>>,
+    ) -> Status,
+    mtftp: unsafe extern "efiapi" fn(
+        this: &Self,
+        operation: TftpOpcode,
+        buffer: Option<NonNull<c_void>>,
```

Comment:
Using patterns like `Option<NonNull<...>>` can be helpful for the public Rust interface, but for the FFI interface I think it's generally clearer to stick close to C definition. So rather than use `Option<NonNull<c_void>>` here, use `*mut c_void` instead. Similar for the `filename` arg.

This makes the calling code simpler too, for example you can replace this:
```rust
let filename = NonNull::from(&filename.to_bytes_with_nul()[0]).cast();
```
With this:
```rust
let filename = filename.as_ptr();
```

Reply:
What about arguments whose type is only a `NonNull<...>` without `Option<...>`:
https://github.com/rust-osdev/uefi-rs/blob/bd7c3cbecf138bfe83d54170f2845d4838d4de41/src/proto/network/pxe.rs#L56
Would you prefer raw pointers in those cases too?

Reply:
Yes, I think on the whole I think I prefer a raw pointer there too.